### PR TITLE
php: Fix php pcre by using external lib

### LINKED ISSUE #31451: mod_php pcre groups are broken with pcre.jit = true
## Issue description

    <?php
      preg_match('/(42)/', '42', $result);
      var_dump($result);
    ?>

Should dump, among other things, "42", but instead results in no match at all. This appears to happen whenever groups are used at all, but simple matches work (eg. '/\d\d/' or simply '/42/' will match '42')

This seems quite severe.

### Steps to reproduce

I've prepared a [nixos-test in a gist](https://gist.github.com/srhb/2d9f5354cdf16c96ec3356f4c5383303). It should fail when `pcre.jit` is set to `true` (the bug) and succeed when set to `false`.

## Technical details

* System: 18.03.git.292f605 (Impala)
* Nix version: 1.11.15
* Nixpkgs version: 292f605 (nixos-unstable)
* Sandboxing enabled: true

### LINKED ISSUE #31499: Broken PCRE JIT in PHP SAPI module for Apache
**TL;DR:** When using PHP 7.1 or lower in conjunction with the Apache SAPI, the `preg_*` functions no longer work as expected.

I stumbled on this while investigating a test failure for nixcloud and I bisected this to the update of the PCRE library to version 8.41 in 119ebee6a4f66d4e623e058e897bdd5928ba478b (#30963). After a bit of debugging and another bisect against the PCRE library I found the commit which is causing this:

https://vcs.pcre.org/pcre?view=revision&revision=1690

With the CLI version of PHP this problem isn't showing up at all, which I found a bit odd.

So I dug a bit deeper and found that PHP bundles PCRE 8.40, which of course doesn't include the PCRE commit mentioned above. Now PHP itself doesn't have `pcre` in its `buildInputs` but Apache uses it.

My suspicion is that some parts of the shared library from Apache are re-used in PHP, but for now I don't have time to dig further into that.

Here a small test which should help to reproduce the problem:

```nix
{ jit ? true, ... }@args:

import <nixpkgs/nixos/tests/make-test.nix> {
  name = "apache-php-sapi-pcre-issue";

  machine = { lib, pkgs, ... }: {
    services.httpd.enable = true;
    services.httpd.adminAddr = "foo@example.org";
    services.httpd.enablePHP = true;
    services.httpd.phpOptions = lib.optionalString (!jit) "pcre.jit=0";
    services.httpd.documentRoot = pkgs.writeTextFile {
      name = "docroot";
      destination = "/index.php";
      text = ''
        <?php
        if (preg_match('/^(?:[0-9]+|[a-z])$/', $_GET['input']))
          echo "YAY\n";
      '';
    };
  };

  testScript = ''
    sub testInput {
        return 'curl "http://localhost/index.php?input='.$_[0].'"'
             . ' | grep -qF YAY';
    }

    $machine->waitForUnit('multi-user.target');
    $machine->succeed(testInput 10);
    $machine->fail(testInput 'foo');
    $machine->succeed(testInput 'x');
  '';
} args
```

Whenever running with `--arg jit false` or the commit before 119ebee6a4f66d4e623e058e897bdd5928ba478b, the test runs fine.

With the `preg_*` functions not working as expected (due to the top-down stack in the JIT), I'd consider this quite risky for people running various PHP web applications that rely on regular expressions to do their input validation.

Maybe it makes sense to disable the PCRE JIT for PHP versions <= 7.1 until we have found out what's wrong?

PHP version 7.2 bundles PCRE 8.41 and doesn't have this problem.

Cc: @qknight, @fpletz, @globin, @grahamc

### PR DESCRIPTION
**Pull request by srhb**
###### Motivation for this change
Fixes #31451 
Fixes #31499

###### Things done
Tested using https://gist.github.com/srhb/2d9f5354cdf16c96ec3356f4c5383303

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by srhb** [comment]
Alternative candidate for the fix: #31525

**Comment by srhb** [comment]
@vcunat is the only one who has voiced their opinion so far, preferring using the external pcre library (which I'm leaning towards as well.)

I've added a NixOS test that blocks the channel advancement (I think?) if it fails that tests that similar issues don't arise again. I could use a review on this, since I'm new to messing with the release files. :)

**Comment by vcunat** [comment]
This should block the big channels and not the small ones.  Given that we do have `php` in the small ones as well, I suggest to make them depend on this check as well (better confirmed by someone else as well).

File: pkgs/development/interpreters/php/default.nix [hunk 1]
```diff
@@ -91,6 +91,12 @@ let
           configureFlags = [ "--enable-pcntl" ];
         };
 
+        # pcre functionality is tested in nixos/tests/php-pcre.nix
```
Comment (orivej) [inline pkgs/development/interpreters/php/default.nix]: It will be more noticeable to say at the top of the file that there is nixos php test available.

**Comment by orivej** [comment]
I agree with @vcunat that we should use external pcre, both because unbundling is more maintanable and because the other approach as it is implemented in php is fragile. It bothers me that you add `pcreExternal` option, whereas we won't support it being off unless we merge #31525 too, but I suppose we can leave it hoping that no one will use it without a good reason.

**Comment by srhb** [comment]
@orivej That is a good point. I can simply set it to `true` instead, but is there another way you would prefer?

**Comment by srhb** [comment]
@orivej How about 7e17685 ?

**Comment by orivej** [comment]
Thank you.

**Comment by orivej** [comment]
Released in b76e7f83de5061d3959887113afb4ed691238635

**Pull request merged**